### PR TITLE
Avoid a constellation roundtrip when a same-origin iframe finishes loading

### LINKED ISSUE #16108: Make a fast path for same-origin iframes to notify parent when loaded
Right now, DocumentProgressHandler::dispatch_load will dispatch the load event for a document, then call Document::notify_constellation_load. This sends a message to the constellation, which in Constellation::handle_subframe_loaded sends a message the event loop for the parent browsing context (if any) of the document that just loaded. In the same-origin case, this is the same event loop that started the process. I propose that DocumentProgressHandler check whether the parent document exists and is same-origin, and if so notify it - not sure if synchronous or via enqueuing a task is better here. Additionally, we still need to talk to the constellation, but we can send a flag indicating that we've already notified them about the child document having been marked as loaded.

This should improve some of our tests that timeout under load.

### PR DESCRIPTION
**Pull request by jdm**
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #16108 
- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16109)

### CODE REVIEW

**Comment by jdm** [comment]
r? @asajeffrey 

I'm not totally convinced about these changes, but they should make some of our tests that use multiple iframes run better when under load.

File: components/script_traits/script_msg.rs [hunk 1]
```diff
@@ -93,8 +93,9 @@ pub enum ScriptMsg {
     /// <head> tag finished parsing
     HeadParsed,
     /// All pending loads are complete, and the `load` event for this pipeline
-    /// has been dispatched.
-    LoadComplete(PipelineId),
+    /// has been dispatched. The boolean indicates whether the parent document
+    /// has been notified already.
+    LoadComplete(PipelineId, bool),
```
Comment (nox) [inline components/script_traits/script_msg.rs]: Booleans are the devil, could you please use a struct variant?

**Comment by asajeffrey** [comment]
Would we be better off keeping track of the state of whether the notification to the parent has already been sent or not? Then the script thread being notified of loading would be idempotent, and we could simplify the constellation code. We would still be sending a pointless notification message from the constellation to script, so there's an efficiency vs code simplicity trade-off here.

Apart from that, and the `bool` vs enumerated type that @nox raised, lgtm, and yay for less round-tripping!

**Comment by jdm** [comment]
It took me a lot of investigation to finally figure out the issue, and it's made me a bit nervous! Previously I was retrieving the frame element from the browsing context, then calling `global()` on that to get a global that would give me the pipeline for the page containing the iframe. Turns out this is broken when the iframe is adopted into a different global, since the result of `global()` remains the creator, not the window for the document that it is now a part of. I fear that we may have a lot of these subtle bugs lurking :(

**Comment by asajeffrey** [comment]
Yes, I saw that, I didn't realize that `node.global()` gives the originating global, not the global of the current window. I also suspect this is a good source of bugs.

**Comment by asajeffrey** [comment]
@jdm: I think the only remaining issue is the bool which should be an enum?

**Comment by jdm** [comment]
@asajeffrey I don't see a way to only store the information in the script thread without risking races.

**Comment by jdm** [comment]
The failures here come from a timing issue - if the script thread decides that a document is loaded before the constellation's document activity level notification is received, attempts to interact with the iframe's content will use the pre-navigation frame contents. I'm going to give up on this attempted optimization; I think it's going to be too difficult to maintain as part of the ever-more-complex navigation infrastructure.

**Pull request closed without merging**